### PR TITLE
Feature/prediction classification and detection

### PR DESCRIPTION
**Pull request by lumurillo**
Add decorator class for the prediction and two specializations: The classification prediction and the detection prediction.

UPDATE 22/09/2020: The following changes were added:

- Remove the decorator pattern implementation of the Predict class.
- Added a set of Inference Output classes that will be used to display the predicted data.

### CODE REVIEW

File: r2i/classification.cc [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+                                       labels) {
+  this->labels = labels;
+  return RuntimeError();
+}
+
+std::vector< ClassificationInstance > Classification::GetLabels() {
+  return this->labels;
+}
+
+ClassificationInstance Classification::GetRank1() {
```
Comment (jafet-chaves) [inline r2i/classification.cc:34]: I think this should be removed. This a PostProcessing role.

**Review by jafet-chaves — CHANGES_REQUESTED** [review]

File: r2i/classification.cc (hunk 1 above)
Comment (lumurillo) [reply r2i/classification.cc:34]: Once you have the vector of tuples, this is a repetitive logic that can be factorized in here.

**Review by jafet-chaves — CHANGES_REQUESTED** [review]

File: r2i/inferenceoutput.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ *
+ * The contents of this software are proprietary and confidential to RidgeRun,
+ * LLC.  No part of this program may be photocopied, reproduced or translated
+ * into another programming language without prior written consent of
+ * RidgeRun, LLC.  The user is free to modify the source code after obtaining
+ * a software license from RidgeRun.  All source code changes must be provided
+ * back to RidgeRun without any encumbrance.
+*/
+
+#ifndef R2I_INFERENCE_OUTOUT_H
```
Comment (jafet-chaves) [inline r2i/inferenceoutput.h:12]: Very important typo here. Please fix.

File: r2i/detection.cc [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+                                      detections) {
+  this->detections = detections;
+  return RuntimeError();
+}
+
+std::vector< DetectionInstance > Detection::GetDetections() {
+  return this->detections;
+}
+
+} // namespace r2
```
Comment (jafet-chaves) [inline r2i/detection.cc:34]: r2i

**Review by jafet-chaves — CHANGES_REQUESTED** [review]

File: r2i/classification.cc [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+Classification::Classification() {
+  this->type = InferenceOutputType::CLASSIFICATION;
+}
+
+Classification::~Classification() {
+  this->labels.clear();
+}
+
+RuntimeError Classification::SetLabels(std::vector< ClassificationInstance >
```
Comment (michaelgruner) [inline r2i/classification.cc:24]: I think we should make these const

File: r2i/detection.cc [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+Detection::Detection() {
+  this->type = InferenceOutputType::OBJECT_DETECTION;
+}
+
+Detection::~Detection() {
+  this->detections.clear();
+}
+
+RuntimeError Detection::SetDetections(std::vector< DetectionInstance >
```
Comment (michaelgruner) [inline r2i/detection.cc:24]: I think we should make these const

**Review by michaelgruner — APPROVED** [review]

File: r2i/classification.h [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+#define R2I_CLASSIFICATION_H
+
+#include <r2i/inferenceoutput.h>
+
+#include <vector>
+#include <tuple>
+
+#include <r2i/runtimeerror.h>
+
+#define CLASS_LABEL_INDEX 0
```
[the remark this answers is not in the document]
Comment (lumurillo) [reply r2i/classification.h:22]: No needed. I removed it.

File: r2i/classification.h [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+ * into another programming language without prior written consent of
+ * RidgeRun, LLC.  The user is free to modify the source code after obtaining
+ * a software license from RidgeRun.  All source code changes must be provided
+ * back to RidgeRun without any encumbrance.
+*/
+
+#ifndef R2I_CLASSIFICATION_H
+#define R2I_CLASSIFICATION_H
+
+#include <r2i/inferenceoutput.h>
```
[the remark this answers is not in the document]
Comment (lumurillo) [reply r2i/classification.h:15]: This standard: https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

**Pull request merged**